### PR TITLE
Make `iterate(::Reverse)` use indexing by default

### LINKED ISSUE #43101: `last` now broken for custom iterators on master
On 1.0 - 1.7, the following is working
```julia
julia> convergents(fmpz[1, 2])
Nemo.ConvergentsIterator{fmpz, fmpq}(fmpz[1, 2])

julia> last(convergents(fmpz[1, 2]))
3//2
```
On recent master, due to https://github.com/JuliaLang/julia/issues/42943, this is broken and we have the following regression:
```julia
ulia> convergents(fmpz[1, 2])
Nemo.ConvergentsIterator{fmpz, fmpq}(fmpz[1, 2])

julia> last(convergents(fmpz[1, 2]))
ERROR: MethodError: no method matching iterate(::Base.Iterators.Reverse{Nemo.ConvergentsIterator{fmpz, fmpq}})
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at ~/repositories/julia/usr/share/julia/base/range.jl:870
  iterate(::Union{LinRange, StepRangeLen}, ::Integer) at ~/repositories/julia/usr/share/julia/base/range.jl:870
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}} at ~/repositories/julia/usr/share/julia/base/dict.jl:695
  ...
Stacktrace:
 [1] first(itr::Base.Iterators.Reverse{Nemo.ConvergentsIterator{fmpz, fmpq}})
   @ Base ./abstractarray.jl:424
 [2] last(a::Nemo.ConvergentsIterator{fmpz, fmpq})
   @ Base ./abstractarray.jl:483
 [3] top-level scope
   @ REPL[5]:1
```
It does not really matter what the type here is. It is a custom iterator satisfying the iterator interface.

Why was `last` broken for iterators? I thought we are not breaking user code (which does not rely on internals) before 2.0 :)

### PR DESCRIPTION
**Pull request by MasonProtter**
With this change we can do things like
```julia
julia> struct MyVector
           v::Vector
       end

julia> Base.getindex(mv::MyVector, args...) = Base.getindex(mv.v, args...)

julia> Base.eachindex(mv::MyVector) = eachindex(mv.v)

julia> for x ∈ Iterators.reverse(MyVector([:a, :b, :c]))
           @show x
       end
x = :c
x = :b
x = :a
```

I'm proposing this as an alternative to https://github.com/JuliaLang/julia/pull/43106 in order to fix https://github.com/JuliaLang/julia/issues/43101

### CODE REVIEW

**Comment by MasonProtter** [comment]
I think the most significant downside I can forsee of this is that if one has a type that does not support indexing, then you may get worse error messages than before. 

Compare (old)
```julia
julia> struct Foo
           a::Int
           b::Int
       end

julia> for x ∈ Iterators.reverse(Foo(1, 2))
           @show x
       end
ERROR: MethodError: no method matching iterate(::Base.Iterators.Reverse{Foo})
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at ~/julia/usr/share/julia/base/range.jl:821
  iterate(::Union{LinRange, StepRangeLen}, ::Integer) at ~/julia/usr/share/julia/base/range.jl:821
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}} at ~/julia/usr/share/julia/base/dict.jl:695
  ...
Stacktrace:
 [1] top-level scope
   @ ./REPL[2]:1
```
against (new)
```julia
julia> struct Foo
           a::Int
           b::Int
       end

julia> for x ∈ Iterators.reverse(Foo(1, 2))
           @show x
       end
ERROR: MethodError: no method matching keys(::Foo)
Closest candidates are:
  keys(::IndexStyle, ::AbstractArray, ::AbstractArray...) at ~/julia/usr/share/julia/base/abstractarray.jl:350
  keys(::Tuple) at ~/julia/usr/share/julia/base/tuple.jl:72
  keys(::Tuple, ::Tuple...) at ~/julia/usr/share/julia/base/tuple.jl:77
  ...
Stacktrace:
 [1] eachindex(itrs::Foo)
   @ Base ./abstractarray.jl:276
 [2] iterate(A::Base.Iterators.Reverse{Foo})
   @ Base.Iterators ./REPL[1]:2
 [3] top-level scope
   @ ./REPL[7]:1
```

**Comment by vtjnash** [comment]
I think we should do this _in addition_ to https://github.com/JuliaLang/julia/pull/43106

**Comment by Seelengrab** [comment]
Since `Iterators.reverse` is a lazy wrapper anyway, we could check `isapplicable` on construction and manually throw a `MethodError`, right? Wouldn't be pretty, but the construction itself is probably not performance critical and it wouldn't be the first time a custom `MethodError` is thrown for clarity sake.

File: base/iterators.jl [hunk 1]
```diff
@@ -112,6 +112,15 @@ last(r::Reverse) = first(r.itr) # the first shall be last
     (A.itr[idx], (state[1], itrs))
 end
 
+# Fallback method of `iterate(::Reverse{T})` which assumes the collection has `getindex(::T) and `reverse(eachindex(::T))`
+# don't propagate inbounds for this just in case
+function iterate(A::Reverse state=(reverse(eachindex(A.itr)),))
```
[the remark this answers is not in the document]
Comment (MasonProtter) [reply base/iterators.jl:117]: Maybe use `Base.reverse`?

**Comment by MasonProtter** [comment]
> I think we should do this _in addition_ to #43106

But #43106 makes `last` use indexing instead of `reverse` by default..

**Comment by JeffBezanson** [comment]
Our thinking is that implementing `last` is easier (and could be implemented for strictly more types) than implementing reverse iteration, so we might as well have both fall back to indexing, and types with reverse iteration should just add `last` methods.

**Pull request merged**